drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,965 @@
+{
+	int ret = 0;
+	void *buffer;
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->tx;
+	I2S_Type *base = cfg->base;
+
+	/* retrieve buffer from input queue */
+	ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("No buffer in input queue to start transmission");
+		return ret;
+	}
+
+	i2s_mcux_config_dma_blocks(dev, I2S_DIR_TX, (uint32_t *)buffer);
+
+	LOG_DBG("dma_slot is %d", stream->dma_cfg.dma_slot);
+	LOG_DBG("channel_direction is %d", stream->dma_cfg.channel_direction);
+	LOG_DBG("complete_callback_en is %d",
+		stream->dma_cfg.complete_callback_en);
+	LOG_DBG("error_callback_en is %d", stream->dma_cfg.error_callback_en);
+	LOG_DBG("source_handshake is %d", stream->dma_cfg.source_handshake);
+	LOG_DBG("dest_handshake is %d", stream->dma_cfg.dest_handshake);
+	LOG_DBG("channel_priority is %d", stream->dma_cfg.channel_priority);
+	LOG_DBG("source_chaining_en is %d", stream->dma_cfg.source_chaining_en);
+	LOG_DBG("dest_chaining_en is %d", stream->dma_cfg.dest_chaining_en);
+	LOG_DBG("linked_channel is %d", stream->dma_cfg.linked_channel);
+	LOG_DBG("source_data_size is %d", stream->dma_cfg.source_data_size);
+	LOG_DBG("dest_data_size is %d", stream->dma_cfg.dest_data_size);
+	LOG_DBG("source_burst_length is %d", stream->dma_cfg.source_burst_length);
+	LOG_DBG("dest_burst_length is %d", stream->dma_cfg.dest_burst_length);
+	LOG_DBG("block_count is %d", stream->dma_cfg.block_count);
+
+	dma_config(stream->dev_dma, stream->channel, &stream->dma_cfg);
+
+	/* put buffer in output queue */
+	ret = k_msgq_put(&stream->out_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("failed to put buffer in output queue");
+		return ret;
+	}
+
+	/* Enable TX DMA */
+	base->FIFOCFG |= I2S_FIFOCFG_DMATX_MASK;
+
+	ret = dma_start(stream->dev_dma, stream->channel);
+	if (ret < 0) {
+		LOG_ERR("dma_start failed (%d)", ret);
+		return ret;
+	}
+
+	I2S_Enable(base);
+	I2S_EnableInterrupts(base, (uint32_t)kI2S_TxErrorFlag);
+
+	return 0;
+}
+
+static int i2s_mcux_rx_stream_start(const struct device *dev)
+{
+	int ret = 0;
+	void *buffer[NUM_DMA_BLOCKS];
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->rx;
+	I2S_Type *base = cfg->base;
+	uint8_t num_of_bufs;
+
+	num_of_bufs = k_mem_slab_num_free_get(stream->cfg.mem_slab);
+
+	ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[0],
+						K_NO_WAIT);
+	if (ret != 0) {
+		LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+		return ret;
+	}
+
+	if (num_of_bufs > 1) {
+		/*
+		 * Allocate a second receive buffer from SLAB. This will allow
+		 * us to use a Ping-Pong buffer scheme during receive
+		 */
+		ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer[1],
+							K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+			return ret;
+		}
+	}
```

Comment:
If the DMA engine does not work reliably with only 1 descriptor maybe we should enforce that there has to be at least 2 free memory blocks / 2 descriptors to start RX transfer? It seems that if the DMA engine is started with one descriptor then it will always use one descriptor even if more RX blocks becomes available.

Reply:
Is it a reasonable use case for the user to do RX with a single buffer in the memory slab?

Reply:
No, not really. But if DMA engine holds always 2 RX buffers then the RX buffer pool will have to be a bit larger. Application also needs to 'own' some buffers so it has time to process and free them. There is some cost in requiring 2 free memory blocks / 2 descriptors though reliability is more important. If DMA engine needs 2 buffers to work well we should demand that at least 2 free RX buffers are available when RX transfer is started.